pkg/controller/cluster/main-controller.go

```diff
@@ -344,143 +344,160 @@ func NewController(podName string, namespacesToWatch set.StringSet, kubeClientSe
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.statefulSetListerSynced, c.deploymentListerSynced, c.tenantsSynced, c.policyBindingListerSynced); !ok {
+		panic("failed to wait for caches to sync")
+	}
 
-		select {}
+	klog.Info("Starting workers")
+	// Launch two workers to process Tenant resources
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
 	}
 
-	// runSTS starts the STS API even if the pod is not the leader
-	runSTS := func(ctx context.Context) <-chan interface{} {
-		// stsServerWillStart is a channel for the STS Server API
-		stsServerWillStart := make(chan interface{})
+	// Launch a single worker for Health Check reacting to Pod Changes
+	go wait.Until(c.runHealthCheckWorker, time.Second, stopCh)
+
+	// Launch a goroutine to monitor all Tenants
+	go c.recurrentTenantStatusMonitor(stopCh)
 
+	// 1) we need to make sure we have console TLS certificates (if enabled)
+	if isOperatorConsoleTLS() {
+		klog.Info("Waiting for Console TLS")
+		wg.Add(1)
 		go func() {
-			klog.Infof("Starting STS API server")
-			close(stsServerWillStart)
-			publicCertPath, publicKeyPath := c.waitSTSTLSCert()
-			certsManager, err := xcerts.NewManager(ctx, *publicCertPath, *publicKeyPath, LoadX509KeyPair)
+			defer wg.Done()
+			klog.Infof("Console TLS is enabled, starting console TLS certificate setup")
+			err := c.recreateOperatorConsoleCertsIfRequired(ctx)
 			if err != nil {
-				klog.Infof("STS HTTPS server ListenAndServeTLS failed: %v", err)
 				panic(err)
 			}
-			serverCertsManager = certsManager
-			c.sts.TLSConfig = c.createTLSConfig(serverCertsManager)
-			if err := c.sts.ListenAndServeTLS("", ""); err != http.ErrServerClosed {
-				klog.Infof("STS  HTTPS server ListenAndServeTLS failed: %v", err)
-				panic(err)
+			klog.Infof("Restarting Console pods")
+			err = c.rolloutRestartDeployment(getConsoleDeploymentName())
+			if err != nil {
+				klog.Errorf("Console deployment didn't restart: %s", err)
 			}
 		}()
+	} else {
+		klog.Infof("Console TLS is not enabled")
+	}
 
-		return stsServerWillStart
+	// 2) we need to make sure we have STS API certificates (if enabled)
+	if IsSTSEnabled() {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			klog.Infof("STS is enabled, starting STS API certificate setup")
+			c.generateSTSTLSCert()
+		}()
 	}
 
+	wg.Wait()
+
+	var err error
+	for {
+		select {
+		case err = <-apiServerChannel:
+			if err != http.ErrServerClosed {
+				klog.Errorf("API Server stoped: %v", err)
+				go startAPIServer(ctx, c, apiServerChannel)
+			}
+			// webswerver was instructed to stop, do not attempt to restart
+			continue
+		case err = <-upgradeServerChannel:
+			if err != http.ErrServerClosed {
+				klog.Errorf("Upgrade Server stoped: %v", err)
+				go startUpgradeServer(ctx, c, apiServerChannel)
+			}
+			// webswerver was instructed to stop, do not attempt to restart
+			continue
```

Comment:
We should use a single channel IMO and send the Error typed so that you can detect it and restart relevant servers if needed.

Reply:
addressed the comments @harshavardhana